### DISCUSSION TITLE
GMSK SX1278

### DISCUSSION THREAD

**Discussion by ansocket** [jgromes/RadioLib#644] [Q&A]
Hi there!

Anyone tries to receive GMSK signal on 9600bit/s? I have many high level bits. In screeshot I try to receive packet with some 0x55.
![image](https://user-images.githubusercontent.com/112548510/207818067-76309ff2-5a27-400b-b77e-86cca82d4ae3.png)
 I can see a packet in direct mode on 19200bit/s by connection DIO2 to logic analizator, but interrupt hundler doesn't fire. 

Cheers, Klim

**Comment by jgromes** [comment]
I need more information about your setup. Are you using direct mode for the GMSK? What platform are you using: Arduino, ESP32, ...? What is the signal source, another SX1278? And can you post code for both transmitter and receiver?

**Reply by ansocket** [reply]
1) Yes, I'm using direct mode for output raw data to a logic analizator.
2) STM32F401CC + RA-02 module (sx1278)
3) Signal source - GMSK signal from National Instrument transmitter in Labview.
4) Receiver
```
#include <RadioLib.h>
SX1278 radio = new Module(PC15, PB4, PB3, PB5);
const int pin = PB6;

void setup() {
  Serial.begin(115200);

  Serial.print(F("[SX1278] Initializing ... "));
  int state = radio.beginFSK(434.0, 9.6, 4.8, 50.0, 10, 16, false);
  radio.setDataShaping(RADIOLIB_SHAPING_0_5);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }

  // set the direct mode sync word
  // the default RadioLib FSK sync word is 0x12AD
  // we can add in some preamble bits, to lower the chance
  // of false detection
  radio.setDirectSyncWord(0x555512AD, 32);

  // set function that will be called each time a bit is received
  radio.setDirectAction(readBit);

  // start direct mode reception
  radio.receiveDirect();
}

// this function is called when a new bit is received
void readBit(void) {
  // read the data bit
  radio.readBit(pin);
}

void loop() {
  // we expect the packet to contain the string "Hello World!",
  // a length byte and 2 CRC bytes, that's 15 bytes in total
  if(radio.available() >= 15) {
    Serial.println("[SX1278] Received packet in direct mode!");
    while(radio.available()) {
      // read a byte
      byte b = radio.read();
  
      // print it
      Serial.print(b, HEX);
      Serial.print('\t');
      Serial.write(b);
      Serial.println();
    }
  }
}
```
On the screeshot D2 - DIO2, D3 - DIO0

**Reply by jgromes** [reply]
Are you certain D3 is on DIO0? It seems more like the clock output which should be on DIO1.

**Reply by ansocket** [reply]
Yes, my mistake. Dclk on DIO1.
Tommorow I can take more detailed info about transmissions at my workplace.
Main question: does sx1278 have issues with GMSK on low bitrates? For example, CC1101 cannot work in MSK if bitrate is below 26kbit/s.
I understand, this question is probably more about hardware than software, but if you something about that I would glad to see it.

**Reply by jgromes** [reply]
I'm not aware of any issue on low bitrates. Regarding the data signal, maybe this is caused by frequency error? I don't know which SX1278 module you are using, but they tend to have rather poor crystals. For LoRa it's not an issue because of the very large bandwidth, but for narrowband MSK this could be a problem.

**Reply by ansocket** [reply]
Hmm. I saw on a logic analizator that Dclk is not stable 9600 or 9600.960 (actual BR in datasheet). It can change more than 50kHz down. Can bad crystal oscillator make this? If I increase bitrate to 19200 or more, unstable dclk stays, but reception is much better.

**Reply by jgromes** [reply]
I meant error in the actual carrier frequency of the SX1278 (434 MHz), if it's too low then it would see all the bits as 1 (or as 0 if it's too high).

**Reply by ansocket** [reply]
Ok, I'll measure by osciloscope tommorow. I have all "1" on 2.4kbit/s bitrate...

**Reply by ansocket** [reply]
Hello! The problem was in temperature unstable crystal. When I came in a cold office in the morning, reception was not bad on 9600bit/s. I could see a whole packet, interrupt correctly fired. But while temperature was increasing, reception became worst. So, I decrease output frequency on 2kHz less and reception became perfect again! => I should use AFC to correct a frequency error. Thank you
